taichi/codegen/llvm/codegen_llvm.cpp

```diff
@@ -369,47 +420,102 @@ void TaskCodeGenLLVM::visit(UnaryOpStmt *stmt) {
     llvm::CastInst::CastOps cast_op;
     auto from = stmt->operand->ret_type;
     auto to = stmt->cast_type;
+    TI_ASSERT_INFO(
+        from->is<TensorType>() == to->is<TensorType>(),
+        "Cannot cast between tensor type and non-tensor type: {} v.s. {}",
+        from->to_string(), to->to_string());
     if (from == to) {
       llvm_val[stmt] = llvm_val[stmt->operand];
-    } else if (is_real(from) != is_real(to)) {
-      if (is_real(from) && is_integral(to)) {
-        cast_op = is_signed(to) ? llvm::Instruction::CastOps::FPToSI
-                                : llvm::Instruction::CastOps::FPToUI;
-      } else if (is_integral(from) && is_real(to)) {
-        cast_op = is_signed(from) ? llvm::Instruction::CastOps::SIToFP
-                                  : llvm::Instruction::CastOps::UIToFP;
+    } else if (is_real(from) != is_real(to) ||
```

Comment:
DataType has this method of get_element_type() which returns the primitive dtype.

You can simply check for is_real(from.get_element_type()), cause the branch seems agnostic of whether it's TensorType or not.

Same for most of the checks in visit(BinaryOpStmt*)